spec/integration/parser/compiler_spec.rb

```diff
@@ -522,4 +522,31 @@ class xport_test {
       expect(catalog).to have_resource("Notify[nbr2]").with_parameter(:message, 'overridden')
     end
   end
+
+
+  context "when converting catalog to resource" do
+    it "the same environment is used for compilation as for transformation to resource form" do
+        Puppet[:code] = <<-MANIFEST
+          notify { 'dummy':
+          }
+        MANIFEST
+
+        # Create an instance of a mocha state machine that checks that the environment is
+        # 'production'
+        in_production_state = Class.new do
+
+          def active?
+            Puppet.lookup(:current_environment).name == :production
+          end
+
+          def to_s
+            # displayed if the expectation about this state fails
+            "in production environment"
+          end
+        end.new
```

Comment:
afaikt, the when(state) is the only option to get dynamic code into the equation. Trying to specify it as a mock return and using a block to compute the return did not work at all (it seems to want to bind the value when the mock is created). Maybe something clever can be done with the StateMachine class as opposed to only State... Appreciate you taking a look.